### DISCUSSION TITLE
P192 リスト18.14のコードについての質問

### DISCUSSION THREAD

**Discussion by mizutec** [budougumi0617/go_todo_app#38] [Q&A]
質問失礼いたします。

表題のコード中で**インターフェースが実際の型のシグネチャと一致しているか確認するコード**というものがございますが、これは一体何をしているのでしょうか？

```go
var  (
    _ Beginner = (*sqlx.DB)(nil)
    ...
 )
```

他書籍となりますが「実用Go言語」のP87のブランク識別子を利用したテクニックと同様のことが行われているのだと理解しております。しかし上記のコードの`(*sqlx.DB)(nil)`で何が行われているのかが理解できませんでした。

GoでWebアプリケーションを作ってみようとGoを学び始めたタイミングで、素晴らしい書籍に出会えたことに感謝しております！

**Comment by budougumi0617** [comment]
ご質問ありがとうございます。たしかにトリッキーですよねコレ。


`(*sqlx.DB)(nil)` は変数宣言を使わずに `*sqlx.DB` 型の値を作っています。 `nil` なのでメモリアロケーションなどが発生しません。
Go本体の標準パッケージの実装でもよく使われています。

https://github.com/golang/go/blob/f324355d1f482362b87ec4f95ceac00d4b4de797/src/net/net.go#L714-L717

丁寧に書くとこうなります。
```go
var b *sqlx.DB = nil
var _ Beginner = b
```

`var b *sqlx.DB = &sqlx.DB{}` と書くと使わない値のメモリアロケーションが発生するので `nil` で初期化されたポインタの値を作ります（些細な節約でしょうけども）

もっと段階を踏んで書くとこんな感じです。
https://go.dev/play/p/hIucyFEm5kL
```go
var (
	// ポインタ型の変数の初期値はnil
	b *bytes.Buffer = nil
	_ io.Reader     = b

	// ポインタ型の変数の初期値はnilなので、こうでもよい
	b2 *bytes.Buffer
	_  io.Reader = b2

	// 変数bや変数b2は使わないので一行にまとめてしまえばよい	
	_ io.Reader = (*bytes.Buffer)(nil)

	// この書き方は利用されない値のメモリアロケーションが発生する
	// （もしかしたらコンパイラが最適化してアロケーションしていないかもしれないが
	_ io.Reader = &bytes.Buffer{}
)
```

> 「実用Go言語」のP87のブランク識別子を利用したテクニックと同様のことが行われているのだと理解

上記の説明のとおりなので同様の目的の実装をしているものの、もうすこし最適化されている感じですね。

これで説明になっているでしょうか？

**Reply by mizutec** [reply]
ご回答ありがとうございます！
```go
var b *sqlx.DB = nil
var _ Beginner = b
```
難しく考えておりましたがこちらのサンプルで理解できました。

また`(*sqlx.DB)(nil)`が型変換だろうとは考えていたのですが、なぜ`*sqlx.DB`を`()`で括っているのか悩んでしまいましたが、`*sqlx.DB(null)`と書いてしまうとエラーになってしまうのを防止するためという理解でよろしいでしょうか？
Playgroundで試したところポインター型の`*sqlx.DB`ではなく`sqlx.DB`へ型変換しようとしているエラーを確認いたしました。

nilまわり、基礎文法への理解が弱いことが分かりましたので、書籍の中で紹介されている参考文献等も適宜確認しながら引き続き進めさせていただきます。

**Reply by budougumi0617** [reply]
はい、そうです。`(*sqlx.DB)(nil)`ではなく `*sqlx.DB(null)` と書いてしまうと `*(sqlx.DB(null))`と解釈されてキャストエラーになるので括弧を使っています！
